include/contractor/contractor.hpp

```diff
@@ -718,6 +718,50 @@ class Contractor
     }
 
   private:
+    inline void RelaxNode(const NodeID node,
+                          const NodeID forbidden_node,
+                          const int distance,
+                          ContractorHeap &heap)
+    {
+        const short current_hop = heap.GetData(node).hop + 1;
+        for (auto edge : contractor_graph->GetAdjacentEdgeRange(node))
+        {
+            const ContractorEdgeData &data = contractor_graph->GetEdgeData(edge);
+            if (!data.forward)
+            {
+                continue;
+            }
+            const NodeID to = contractor_graph->GetTarget(edge);
+            if (forbidden_node == to)
+            {
+                continue;
+            }
+            const int to_distance = distance + data.distance;
+
+            // New Node discovered -> Add to Heap + Node Info Storage
+            if (!heap.WasInserted(to))
+            {
+                heap.Insert(to, to_distance, ContractorHeapData(current_hop, false));
+            }
+            // Found a shorter Path -> Update distance
+            else if (to_distance < heap.GetKey(to))
+            {
+                heap.DecreaseKey(to, to_distance);
+                heap.GetData(to).hop = current_hop;
+            }
+        }
+    }
+
+    inline int FindSelfLoop(const NodeID node, int max_distance, ContractorThreadData *const data)
+    {
+        ContractorHeap &heap = data->heap;
+        heap.Clear();
+        heap.Insert(node, INT_MAX, ContractorHeapData(0, true)); // mark the node itself as a target
```

Comment:
Hmm int max here depends on the value type used for representation.. this is kind of ugly :/